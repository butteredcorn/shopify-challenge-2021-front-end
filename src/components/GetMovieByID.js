import { useState, useEffect } from 'react'
import '../styles/GetMovieByID.css'
import styled from 'styled-components'

const MovieContainer = styled.div`
    background-color: ${props => props.theme.bg.secondary};
    color: ${props => props.theme.text.tertiary};
    border: 1px solid ${props => props.theme.bg.primary};
`

const GetMoviesByID = ({error, loading, movie, nominate, remove}) => {
    const [nominated, setNominated] = useState(false)

    useEffect(() => {
        if (movie.imdbID) {
            const nomination = JSON.parse(localStorage.getItem("nominations")).filter(n => n.imdbID === movie.imdbID)
            if (nomination.length > 0) {
                setNominated(true)
            } else {
                setNominated(false)
            }
        }
    }, [movie.imdbID])

    if (loading) return null;
    if (error) return `${error}.`;

    return (
        <MovieContainer className="container">
            {movie && 
            <div>
                <h2 className="movie-title">{movie.Title} ({movie.Year})</h2>
                <div className="movie-container">
                    <div>
                        <img className="poster" src={movie.Poster} alt="movie poster"/>
                    </div>
                    <div className="movie-details">
                        <span><strong>Genre:</strong> {movie.Genre}</span>
                        <span><strong>Rated:</strong> {movie.Rated}</span>
                        <span><strong>Runtime:</strong> {movie.Runtime}</span>
                        <span><strong>Director:</strong> {movie.Director}</span>
                        <span><strong>Actors:</strong> {movie.Actors}</span>
                        <span><strong>Language:</strong> {movie.Language}</span>
                        <span><strong>Metascore:</strong> {movie.Metascore}</span>
                        <span><strong>IMDB Rating:</strong> {movie.imdbRating}</span>
                        {!nominated && <span><button onClick={() => { 
                            nominate({Title: movie.Title, Year: movie.Year, imdbID: movie.imdbID}) 
                            setNominated(true)
                        }}>Nominate</button></span>}
                        {nominated && <span><button onClick={() => { 
                            remove(movie.imdbID) 
                            setNominated(false)
                        }}>Remove</button></span>}
                    </div>
                </div>
            </div>}
        </MovieContainer>
    )
}

export default GetMoviesByID;