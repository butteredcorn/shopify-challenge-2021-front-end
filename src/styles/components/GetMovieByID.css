.movie-container {
    padding: 1.5rem;
}

.movie-title {
    font-size: 2.5rem;
    font-weight: bold;
    margin-top: 0rem;
    margin-bottom: 1rem;
}

.movie-detail-container {
    display: flex;
    flex-direction: row;
}

.poster-container {
    margin-right: 1.5rem;
}

.poster {
    margin-right: 0;
    max-width: 20rem;
}


.movie-details {
    display: flex;
    flex-direction: column;
    font-size: 1.2rem;
}

.movie-details > * {
    padding-bottom: 0.5rem;
    display: flex;
}

strong {
    min-width: 8rem;
}

@media only screen and (max-width: 600px) {
    
    .poster-container {
        margin: 0;
    }

    .poster {
        margin-right: 0;
        width: 100%;
    }

    .movie-container {
        flex-direction: column;
    }

    .movie-title {
        font-size: 1.8rem;
    }
}