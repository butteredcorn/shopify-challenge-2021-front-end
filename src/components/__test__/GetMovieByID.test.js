import { render, screen, cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer'
import { BrowserRouter } from 'react-router-dom'
import ThemesProvider from '../../providers/ThemeProvider'
import GetMovieByID from '../GetMovieByID';

afterEach(cleanup);

class LocalStorageMock {
    constructor() {
      this.store = {};
    }
  
    clear() {
      this.store = {};
    }
  
    getItem(key) {
      return this.store[key] || null;
    }
  
    setItem(key, value) {
      this.store[key] = String(value);
    }
  
    removeItem(key) {
      delete this.store[key];
    }
  };
  
global.localStorage = new LocalStorageMock;
localStorage.setItem("nominations", JSON.stringify([]));

it('renders GetMovieByID component with "Shopify Challenge" if prop "movie" is set to movie object.', () => {
  render(<ThemesProvider><GetMovieByID movie={{Title: "Shopify Challenge", Year: "2021", Genre: "Front End Developement", Rated: "E for Everyone", Runtime: "Life long journey", Director: "Justin Yee", Actors: "Justin Yee", Language: "Javascript, React, React Native, Node, C#, PHP, Python, Java, and more.", Metascore: "96", imdbRating: "9.6/10", imdbID: "#1"}} /></ThemesProvider>);
  const message = screen.getByText(/Shopify Challenge/i);
  expect(message).toBeInTheDocument();
});

it('renders GetMovieByID component with "Loading..." if prop "loading" is set to true', () => {
    render(<BrowserRouter><ThemesProvider><GetMovieByID loading={true} movie={{Title: "Shopify Challenge", Year: "2021", Genre: "Front End Developement", Rated: "E for Everyone", Runtime: "Life long journey", Director: "Justin Yee", Actors: "Justin Yee", Language: "Javascript, React, React Native, Node, C#, PHP, Python, Java, and more.", Metascore: "96", imdbRating: "9.6/10", imdbID: "#1"}} /></ThemesProvider></BrowserRouter>);
    const message = screen.queryByText(/Loading.../i);
    expect(message).toBeInTheDocument();
});

it('expect GetMovieByID to render an error if an error is passed in as a prop', () => {
    render(<ThemesProvider><GetMovieByID error={new Error("Hello World!")} movie={{Title: "Shopify Challenge", Year: "2021", Genre: "Front End Developement", Rated: "E for Everyone", Runtime: "Life long journey", Director: "Justin Yee", Actors: "Justin Yee", Language: "Javascript, React, React Native, Node, C#, PHP, Python, Java, and more.", Metascore: "96", imdbRating: "9.6/10", imdbID: "#1"}} /></ThemesProvider>);
    const message = screen.getByText(/Hello World!/i);
    expect(message).toBeInTheDocument();
});

it('matches a loading snapshot', () => {
    const json = renderer.create(<ThemesProvider><GetMovieByID loading={true} movie={{Title: "Shopify Challenge", Year: "2021", Genre: "Front End Developement", Rated: "E for Everyone", Runtime: "Life long journey", Director: "Justin Yee", Actors: "Justin Yee", Language: "Javascript, React, React Native, Node, C#, PHP, Python, Java, and more.", Metascore: "96", imdbRating: "9.6/10", imdbID: "#1"}} /></ThemesProvider>).toJSON();
    expect(json).toMatchSnapshot();
});

it('matches a snapshot', () => {
  const json = renderer.create(<BrowserRouter><ThemesProvider><GetMovieByID movie={{Title: "Shopify Challenge", Year: "2021", Genre: "Front End Developement", Rated: "E for Everyone", Runtime: "Life long journey", Director: "Justin Yee", Actors: "Justin Yee", Language: "Javascript, React, React Native, Node, C#, PHP, Python, Java, and more.", Metascore: "96", imdbRating: "9.6/10", imdbID: "#1"}} /></ThemesProvider></BrowserRouter>).toJSON();
  expect(json).toMatchSnapshot();
});