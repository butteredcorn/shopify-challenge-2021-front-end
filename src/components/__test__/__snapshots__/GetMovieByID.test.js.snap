// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches a loading snapshot 1`] = `
<div
  className="sc-gsDJrp jeaySH movie-container"
>
  <p>
    Loading ...
  </p>
</div>
`;

exports[`matches a snapshot 1`] = `
<div
  className="sc-gsDJrp jeaySH movie-container"
>
  <div>
    <h2
      className="sc-dkPtyc iJlIGG movie-title"
    >
      Shopify Challenge
       (
      2021
      )
    </h2>
    <div
      className="movie-detail-container"
    >
      <div
        className="poster-container"
      >
        <img
          alt="movie poster"
          className="poster"
        />
      </div>
      <div
        className="movie-details"
      >
        <span>
          <strong>
            Genre:
          </strong>
           
          Front End Developement
        </span>
        <span>
          <strong>
            Rated:
          </strong>
           
          E for Everyone
        </span>
        <span>
          <strong>
            Runtime:
          </strong>
           
          Life long journey
        </span>
        <span>
          <strong>
            Director:
          </strong>
           
          Justin Yee
        </span>
        <span>
          <strong>
            Actors:
          </strong>
           
          Justin Yee
        </span>
        <span>
          <strong>
            Language:
          </strong>
           
          Javascript, React, React Native, Node, C#, PHP, Python, Java, and more.
        </span>
        <span>
          <strong>
            Metascore:
          </strong>
           
          96
        </span>
        <span>
          <strong>
            IMDB Rating:
          </strong>
           
          9.6/10
        </span>
        <span>
          <button
            onClick={[Function]}
          >
            Nominate
          </button>
        </span>
      </div>
    </div>
  </div>
</div>
`;
